src/serializer/functions.js

```diff
@@ -263,36 +280,35 @@ export class Functions {
       );
       invariant(additionalFunctionEffects);
       effects = additionalFunctionEffects.effects;
+      if (this.writeEffects.has(functionValue)) {
+        let error = new CompilerDiagnostic(
+          "Trying to optimize a function with two parent optimized functions, which is not currently allowed.",
+          functionValue.expressionLocation,
+          "PP1009",
+          "FatalError"
+        );
+        if (realm.handleError(error) !== "Recover") throw new FatalError();
+      }
       this.writeEffects.set(functionValue, additionalFunctionEffects);
 
-      // look for newly registered optimized functions
-      let modifiedProperties = additionalFunctionEffects.effects.modifiedProperties;
       // Conceptually this will ensure that the nested additional function is defined
       // although for later cases, we'll apply the effects of the parents only.
       this.realm.withEffectsAppliedInGlobalEnv(() => {
-        for (let [propertyBinding] of modifiedProperties) {
-          let descriptor = propertyBinding.descriptor;
-          if (descriptor && propertyBinding.object === optimizedFunctionsObject) {
-            let newValue = descriptor.value;
-            invariant(newValue instanceof Value); //todo: this does not seem invariantly true
-            let newEntry = this.__optimizedFunctionEntryOfValue(newValue);
-            if (newEntry) {
-              additionalFunctions.add(newEntry.value);
-              getEffectsFromAdditionalFunctionAndNestedFunctions(newEntry.value, newEntry.argModelString);
-              // Now we have to rember the stack of effects that need to be applied to deal with
-              // this additional function.
-            }
-          }
+        let newOptFuncs = this._generateOptimizedFunctionsFromRealm();
+        for (let newEntry of newOptFuncs) {
+          additionalFunctions.add(newEntry.value);
+          this._withEmptyOptimizedFunctionList(newEntry, recordWriteEffectsForOptimizedFunctionAndNestedFunctions);
         }
+        // Now we have to remember the stack of effects that need to be applied to deal with
+        // this additional function.
         return null;
       }, additionalFunctionEffects.effects);
       invariant(additionalFunctionStack.pop() === functionValue);
-      for (let t2 of this.realm.tracers) t2.endOptimizingFunction(currentOptimizedFunctionId);
     };
 
     while (additionalFunctionsToProcess.length > 0) {
       let funcObject = additionalFunctionsToProcess.shift();
```

Comment:
This fancy looking while look with `shift` is still just a simple iteration, right? Why not write a `for(x of ...)` loop? Less text to read.